### PR TITLE
fix issue #122

### PR DESCRIPTION
**Pull request by scigghia**
simple fix for https://github.com/OCA/OCB/issues/122

### CODE REVIEW

File: addons/product/product.py [hunk 1]
```diff
@@ -685,9 +685,9 @@ def name_search(self, cr, user, name='', args=None, operator='ilike', context=No
             positive_operators = ['=', 'ilike', '=ilike', 'like', '=like']
             ids = []
             if operator in positive_operators:
-                ids = self.search(cr, user, [('default_code','=',name)]+ args, limit=limit, context=context)
+                ids = self.search(cr, user, args + [('default_code','=',name)], limit=limit, context=context)
```
Comment (eLBati) [inline addons/product/product.py:688]: AFAIKS `args` is supposed to be a generic list of tuples in the form accepted by the search method. Adding a tuple before args, like `[('default_code','=',name)]+ args`, means adding that condition with AND.
So, `args + [('default_code','=',name)]` could produce a not valid list of tuples.
I think you should build a correct `args` in your module

**Comment by StefanRijnhart** [comment]
Invalid bug, IMHO. See #122 :-1:

**Comment by gdgellatly** [comment]
I'm 50/50 on it. Product name search is shitty code to work with and i get
the issue and experienced similar myself trying to adapt it, but i dont
think it fits the backports criteria plus to actually "fix" it will be
quite a departure from official if it doesn't land there.

**Comment by OpenCode** [comment]
@eLBati every condition I write in the code is added to existing one as AND parameter. If we move args before the existing list we can manage OR condition, too. Have you different idea to implement this?

@StefanRijnhart for me the bug is valid. The position of args is neutral if yu use AND but is important if you need to use OR.

**Comment by gdgellatly** [comment]
Just looked
at the original bug submission
and the reporters code was attempting to put sql inside domain.  In any
case on product search the first search of the 3 puts the default code
first, but later ones do not, so if the domain isn't working the domain is
wrong.

**Comment by OpenCode** [comment]
The case where default_code isn't the first search is this:

//github.com/OCA/OCB/blob/8.0/addons/product/product.py#L1084

This search return product where default AND name satisfy the condition. I can't add an extra OR operator in my domain otherwise the code before crashes

**Comment by gdgellatly** [comment]
I have no idea what your code is, but generally with a bit of thinking even
this crappy name search can be overcome to get the domain right.  But worst
case nuclear option below.  Set your own flag to override the default name
search, insert name into the arguments and away you go.  To be honest I
don't really understand a scenario where you would want to OR a negative
search term with the same search argument on 3 different fields, because
basically you will get back everything unless the values are already very
similar.

``` python
def name_search(blah blah blah....
    if context and context.get('override_search'):
        if args is None:
            args = []
        #insert name arg wherever into args
        name = ''
    return super(blah blah blah)
```

**Comment by gdgellatly** [comment]
I have no idea what your code is, but generally with a bit of thinking even
this crappy name search can be overcome to get the domain right.  But worst
case nuclear option below.  Set your own flag to override the default name
search, insert name into the arguments and away you go.  To be honest I
don't really understand a scenario where you would want to OR a negative
search term with the same search argument on 3 different fields, because
basically you will get back everything unless the values are already very
similar.

``` python
def name_search(blah blah blah....
    if context and context.get('override_search'):
        if args is None:
            args = []
        #insert name arg wherever into args
        name = ''
    return super(blah blah blah)
```

**Comment by OpenCode** [comment]
@gdgellatly the code is this https://github.com/andreacometa/alternative_codes/blob/7.0/product/product.py#L44

I can't override name value because i need to search in another field, no name.

**Comment by gdgellatly** [comment]
You really need to look at your own code. It has a lot of error potential
and will fail for a lot of reasons but this bug report is not really it.
What if there is one other argument, fail, what if there are no args,
fail, what if the operator is a negative, big fail. Yes your domain and arg
settings are wrong but it is more the whole design is just not right, I'm
sure if you check oca product repo's there will be very similar modules to
what you are trying, product customer code comes to mind as a very similar
concept.  But even if you want to persist with the way you are doing now it
could probably be made to work, just by overriding search instead of name
search and putting these clauses somewhere appropriate notwithstanding the
other issues.  In fact it could probably still be done in name search but
will need a lot more than a few lines.

**Comment by OpenCode** [comment]
1 argument is impossibile because my args list is always added to standard list of tuple in the original name_search function

**Comment by eLBati** [comment]
> @eLBati every condition I write in the code is added to existing one as AND parameter. If we move args before the existing list we can manage OR condition, too. Have you different idea to implement this?

I'm just saying, if args was

```
['|', ('field1', '=', 'value1'), ('field2', '=', value2)]
```

with your code, domain would became

```
['|', ('field1', '=', 'value1'), ('field2', '=', value2), ('default_code','=',name)]
```

that is
(`('field2', '=', value2)` AND `('default_code','=',name)`) OR `('field1', '=', 'value1')`

so, the original purpose of args (`('field1', '=', 'value1')` OR `('field2', '=', value2)`) would be ruined

**Comment by hbrunn** [comment]
:-1: see discussion
